auth/auth.go

```diff
@@ -0,0 +1,281 @@
+}
+
+// Error is an error that occurred during the authentication process.
+type Error struct {
+	// StatusCode is the HTTP status code to respond with.
+	StatusCode int
+	// Cause is the error that caused the failure.
+	Cause error
+}
+
+// Handler is the SAMS authentication handler.
+type Handler struct {
+	config Config
+}
+
+// NewHandler returns a new SAMS authentication handler with the given
+// configuration.
+func NewHandler(config Config) (*Handler, error) {
+	if config.FailureHandler == nil {
+		return nil, errors.New("missing FailureHandler")
+	} else if config.StateSetter == nil {
+		return nil, errors.New("missing StateSetter")
+	} else if config.StateGetter == nil {
+		return nil, errors.New("missing StateGetter")
+	} else if config.StateDeleter == nil {
+		return nil, errors.New("missing StateDeleter")
+	}
+	return &Handler{config: config}, nil
+}
+
+// LoginHandler returns an HTTP handler that redirects the user to the SAMS
+// authentication page.
+//
+// It passes through the "prompt" and "prompt_auth" query parameters to SAMS.
+func (h *Handler) LoginHandler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		// Create a new OIDC provider from the issuer URL using OIDC discovery feature.
+		p, err := oidc.NewProvider(ctx, h.config.Issuer)
+		if err != nil {
+			ctx = WithError(ctx, &Error{
+				StatusCode: http.StatusInternalServerError,
+				Cause:      errors.Wrap(err, "create new provider"),
+			})
+			h.config.FailureHandler.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+
+		// Generate and store a random nonce to the session.
+		nonce, err := randomString()
+		if err != nil {
+			ctx = WithError(ctx, &Error{
+				StatusCode: http.StatusInternalServerError,
+				Cause:      errors.Wrap(err, "generate nonce"),
+			})
+			h.config.FailureHandler.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+		err = h.config.StateSetter(w, r, nonce)
+		if err != nil {
+			ctx = WithError(ctx, &Error{
+				StatusCode: http.StatusInternalServerError,
+				Cause:      errors.Wrap(err, "set state"),
+			})
+			h.config.FailureHandler.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+
+		// NOTE: Using same value for both state and nonce is fine because "state" is
+		// for service self-validation (no CSRF) and "nonce" is for IdP validation (no
+		// repeated attempts).
+		params := url.Values{}
+		params.Add("client_id", h.config.ClientID)
+		params.Add("redirect_uri", h.config.RedirectURI)
+		params.Add("state", nonce)
+		params.Add("nonce", nonce)
+		params.Add("response_type", "code")
+		params.Add("scope", strings.Join(scopes.ToStrings(h.config.RequestScopes), " "))
+
+		// Passthrough the IdP-aware query parameters.
+		params.Add("prompt", r.URL.Query().Get("prompt"))
+		params.Add("prompt_auth", r.URL.Query().Get("prompt_auth"))
+
+		http.Redirect(w, r, p.Endpoint().AuthURL+"?"+params.Encode(), http.StatusFound)
```

Comment:
Generating the URL ourselves via string concatenation seems a tiny bit dubious, but probably alright. But do you think it might be better to set specific fields on a `net.URL` object? Or canonicalize by calling `url.Parse`?

Reply:
🤔 I feel using `net.URL` is >50% likely overkill here, we are only really constructing the query parameter portion.